tls/s2n_connection.c

```diff
@@ -1392,3 +1399,85 @@ int s2n_connection_set_keyshare_by_name_for_testing(struct s2n_connection *conn,
+        *data = NULL;
+    }
+    return S2N_SUCCESS;
+}
+
+static int s2n_sk_X509_pop_free(STACK_OF(X509) **cert_chain) {
+    if (*cert_chain != NULL) {
+        sk_X509_pop_free(*cert_chain, X509_free);
+        *cert_chain = NULL;
+    }
+    return S2N_SUCCESS;
+}
+
+int s2n_connection_get_peer_cert_chain(const struct s2n_connection *conn, struct s2n_cert_chain_and_key *cert_chain_and_key)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(cert_chain_and_key);
+
+    struct s2n_cert_chain *cert_chain = cert_chain_and_key->cert_chain;
+    POSIX_ENSURE_REF(cert_chain);
+    struct s2n_cert **insert = &cert_chain->head;
+    POSIX_ENSURE(*insert == NULL, S2N_ERR_INVALID_CERT_CHAIN);
+
+    const struct s2n_x509_validator *validator = &conn->x509_validator;
+    POSIX_ENSURE_REF(validator);
+    POSIX_ENSURE(validator->state == VALIDATED, S2N_ERR_CERT_NOT_VALIDATED);
+
+    /* X509_STORE_CTX_get1_chain() returns a validated cert chain if a previous call to X509_verify_cert() was successful.
+     * X509_STORE_CTX_get0_chain() is a better API because it doesn't return a copy. But it's not available for Openssl 1.0.2.
+     * See the comments here:
+     * https://www.openssl.org/docs/man1.0.2/man3/X509_STORE_CTX_get1_chain.html
+     */
+    DEFER_CLEANUP(STACK_OF(X509) *cert_chain_validated = X509_STORE_CTX_get1_chain(validator->store_ctx),
+                  s2n_sk_X509_pop_free);
+    POSIX_ENSURE_REF(cert_chain_validated);
+
+    for (size_t cert_idx = 0; cert_idx < sk_X509_num(cert_chain_validated); cert_idx++) {
+        X509 *cert = sk_X509_value(cert_chain_validated, cert_idx);
+        POSIX_ENSURE_REF(cert);
+        DEFER_CLEANUP(uint8_t *cert_data = NULL, s2n_openssl_free);
+        int cert_size = i2d_X509(cert, &cert_data);
+        POSIX_ENSURE_GT(cert_size, 0);
+
+        struct s2n_blob mem = { 0 };
+        if (s2n_alloc(&mem, sizeof(struct s2n_cert)) < S2N_SUCCESS) {
+            goto cleanup;
+        }
+        struct s2n_cert *new_node = (struct s2n_cert *)(void *)mem.data;
+
+        if (s2n_alloc(&new_node->raw, cert_size) < S2N_SUCCESS) {
+            goto cleanup;            
+        }
+        POSIX_CHECKED_MEMCPY(new_node->raw.data, cert_data, cert_size);
+
+        new_node->next = NULL;
+        *insert = new_node;
+        insert = &new_node->next;
+
+    }
+
+    return S2N_SUCCESS;
+    
+    cleanup:
+        /* Walk the chain and free the certs/nodes allocated prior to failure */
+        if (cert_chain) {
+            struct s2n_cert *node = cert_chain->head;
+            while (node) {
+                /* Free the cert */
+                POSIX_GUARD(s2n_free(&node->raw));
+                /* update head so it won't point to freed memory */
+                cert_chain->head = node->next;
+                /* Free the node */
+                POSIX_GUARD(s2n_free_object((uint8_t **)&node, sizeof(struct s2n_cert)));
+                node = cert_chain->head;
+            }
+        }
```

Comment:
This is involved enough to be its own method. You could also do DEFER_CLEANUP then.

Reply:
We can move this to a separate static function, but we can't use DEFER_CLEANUP as the cleanup needs to be done only for failure cases. The memory for the node/certs within the cert_chain need to persist if no failure occurs.  The input `cert_chain_and_key` does not have memory allocated for certs and nodes within it. Ref:

https://github.com/aws/s2n-tls/blob/33b045020a643d8803c5ff827b216315251c8646/crypto/s2n_certificate.c#L172